sklearn/metrics/pairwise.py

```diff
@@ -161,6 +161,128 @@ def check_paired_arrays(X, Y):
 
 
 # Pairwise distances
+def _euclidean_distances_float64(X, Y, Y_norm_squared=None, squared=False,
+                                 X_norm_squared=None):
+    """
+    Compute the euclidean distances between X and Y when both are arrays of
+    float64.
+    """
+    if X_norm_squared is not None:
+        XX = X_norm_squared
+    else:
+        XX = row_norms(X, squared=True)[:, np.newaxis]
+
+    if X is Y:  # shortcut in the common case euclidean_distances(X, X)
+        YY = XX.T
+    elif Y_norm_squared is not None:
+        YY = Y_norm_squared
+    else:
+        YY = row_norms(Y, squared=True)[np.newaxis, :]
+
+    distances = safe_sparse_dot(X, Y.T, dense_output=True)
+    distances *= -2
+    distances += XX
+    distances += YY
+    np.maximum(distances, 0, out=distances)
+
+    if X is Y:
+        # Ensure that distances between vectors and themselves are set to 0.0.
+        # This may not be the case due to floating point rounding errors.
+        distances.flat[::distances.shape[0] + 1] = 0.0
+
+    return distances if squared else np.sqrt(distances, out=distances)
+
+
+def _cast_if_needed(arr, dtype):
+    # This function only exists because the 'astype' method of scipy's sparse
+    # array didn't support the argument 'copy' until version 1.0.
+    if arr.dtype == dtype:
+        return arr
+    return arr.astype(dtype)
+
+
+def _euclidean_distances_cast(X, Y, outdtype, Y_norm_squared=None,
+                              squared=False, X_norm_squared=None):
+    """
+    Compute the euclidean distance matrix between X and Y by casting to float64
+    for a greater numerical stability.
+
+    The computation is done by blocks to limit additional memory usage.
+    """
+    # For performance reasons, swap X and Y if I got X_norm_squared but not
+    # Y_norm_squared
+    if X_norm_squared is not None and Y_norm_squared is None:
+        swap = True
+        X, Y = Y, X
+        X_norm_squared, Y_norm_squared = None, X_norm_squared.T
+    else:
+        swap = False
+
+    # Maximum number of additional bytes to use to case X and Y to float64.
+    maxmem = 10*1024*1024
+    itemsize = np.float64(0).itemsize
+    maxmem /= itemsize
+
+    # Compute the block size that won't use more than maxmem bytes of
+    # additional (temporary) memory.
+    # The amount of additional memory required is:
+    # - a float64 copy of a block of the rows of X if needed;
+    # - a float64 copy of a block of the rows of Y if needed;
+    # - the float64 block result;
+    # - a float64 copy of a block of X_norm_squared if needed;
+    # - a float64 copy of a block of Y_norm_squared if needed.
+    # This is a quadratic equation that we solve to compute the block size that
+    # would use maxmem bytes.
+    XYmem = 0
+    if X.dtype != np.float64:
+        XYmem += X.shape[1]
+    if Y.dtype != np.float64:
+        XYmem += Y.shape[1]
```

Comment:
Perhaps make this X.shape[1] just to reduce reader surprise since they should be equal...

Reply:
On the other hand, it's under a test for `Y.dtype`. So `X.shape[1]` would be equally surprising (at least to me). And this is to account for the memory occupied by the the float64 copy of Y. So semantically, it seems more appropriate to me to use `Y.shape[1]`.
I might add a comment or something to make it more clear.